Issue #5810 - Support bespoke PathSpec for servlets in embedded-jetty

joakime opened the pull request:
Issue #5810
Allows users of embedded-jetty to use ServletContextHandler (and WebAppContext)
to add bespoke `org.eclipse.jetty.http.pathmap.PathSpec` mappings for Servlets.
Works with `WEB-INF/jetty-web.xml` too.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5810 - Support bespoke PathSpec for servlets in embedded-jetty

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -867,6 +898,36 @@ public ServletHolder addServletWithMapping(Class<? extends Servlet> servlet, Str
+            {
+                if (!containsServletHolder(servlet))
+                    setServlets(ArrayUtil.addToArray(holders, servlet, ServletHolder.class));
+            }
```
Might be able to avoid a bit of code duplication if all the add methods eventually called a new method:
```java
public void addServletWithMapping(ServletHolder servlet, ServletMapping mapping)
{
}
```

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -1350,27 +1411,28 @@ protected synchronized void updateMappings()
-            //evaluate path to servlet map based on servlet mappings
-            for (String pathSpec : sms.keySet())
+            // evaluate path to servlet map based on servlet mappings
+            for (String pathSpec : servletUrlMappings.keySet())
```
PathMappings is add order dependent, so I don't really like the fact that we add all the `servletUrlMappings` and them below add all the bespoke path mappings.      Is it possible to have a single loop through `_servletMappings` that adds both standard mappings and bespoke mappings in exactly the order of the mappings themselves?

gregw commented:
On consideration, this is a lot more complex than at first look.
For this to go into a main branch it needs to be consistent and safe.... so that means that we need to consider how additional mapping types affect security.  Eg, if a security constraint is put on a servlet, then it needs to apply to all it's mappings.

This is doable, but it will not be ready in time for 9.4.36.  I believe we need to make a special branch to test this out for a period before merging with the main branch.

gregw committed:
Refactor of fix for #5810

 + Created a common `Mapping` base for `FilterMapping` and `ServletMapping`
 + `Mapping` contains an array of PathSpec.  The `String` based methods assume `ServletPathSpec`.
 + Convert most handling of string pathSpecs to `PathSpec` instances
 + Modified `ConstraintMapping` to also use a PathSpec
 + Modified JSP and annotation handling to cope with any existing non standard path specs
 + Made ServletHandler update mappings respect the order of the mappings

 TODO:
 + Should `ConstraintMapping` extend `Mapping` ?
 + Convert `ConstraintSecurityHandler` to use `PathMappings` so it can handle non standard PathSpecs
 + Review remaining calls to `Mapping.getPathSpecs` and `ConstrainMapping.getPathSpec`
 + Currently Quick Start operates on just the string versions of PathSpec, as it must be able to write a web.xml which can only contain standard path specs.  Should it fail when it sees a non standard pathspec, or assume that it will again be programmatically added during quick start?
 + Review how frequently we are creating new arrays of path spec and/or converting path specs to/from strings.  Best resolution for such things are ultility methods on Mapping (eg `#containsPathSpec`)
 + More tests

gregw commented:
@janbartel Can you please review this closely.  It is a lot more invasive than I hoped, but I still think that we can achieve zero behaviour change if only standard pathspecs are used.   A lot of the changes are good cleanup.   I'm not proposing an early merge for this branch, as I think it will need extensive testing first, specially on the security constraint implications of non standard mappings.

gregw committed:
Refactor of fix for #5810

 + Converted `ConstraintSecurityHandler` to use `PathMappings` so it can handle non standard PathSpecs

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-5810-bespoke-pathspec-support-alt

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Refactor of fix for #5810

 + Fixed test by reverting ServletMapping normalize

gregw committed:
Refactor of fix for #5810

 + More tests

gregw committed:
Updated naming after review

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathSpec.java:
```diff
@@ -85,6 +85,13 @@
     String getSuffix();
 
+    /**
+     *  Test if a normalized string pathspec is equivalent to this pathspec
```
This makes it sound as if the caller is expected to pass in a string that has already had normalize() called on it
```suggestion
     *  Test if a string pathspec is equivalent to this pathspec.
```

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathSpec.java:
```diff
@@ -85,6 +85,13 @@
+    /**
+     *  Test if a normalized string pathspec is equivalent to this pathspec
+     * @param pathSpec String representation to be normalized
+     * @return true of the string represents the same pathspec
```
```suggestion
     * @return true if the string represents the same pathspec
```

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -23,19 +23,22 @@
-            declaration = regex;
+        super(normalize(regex));
+        String declaration = getDeclaration();
         int specLength = declaration.length();
```
Delete specLength variable, only used in 1 place.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -82,13 +85,18 @@ else if (Pattern.matches("^g+l+$", sig))
-        _declaration = declaration;
         _group = group;
         _pathDepth = pathDepth;
         _specLength = specLength;
```
```suggestion
        _specLength = declaration.length();
```

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/AbstractPathSpec.java:
```diff
@@ -22,6 +22,19 @@
+
+    protected AbstractPathSpec(String declaration)
+    {
+        _declaration = declaration;
```
Need a null check on declaration.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -23,19 +23,22 @@
-            declaration = regex.substring("regex|".length());
-        else
-            declaration = regex;
+        super(normalize(regex));
```
Need null protection for regex.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -82,13 +85,18 @@ else if (Pattern.matches("^g+l+$", sig))
+    @Override
+    public boolean is(String pathSpec)
+    {
+        return getDeclaration().equals(normalize(pathSpec));
```
Unless you ensure declaration can never be null, this can throw NPE.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -26,17 +26,16 @@
-     * If a servlet or filter path mapping isn't a suffix mapping, ensure
-     * it starts with '/'
-     *
+     * Normalize servlet path spec:<ul>
```
Ideally, this class would have been called UrlPatternPathSpec - servlets, filters and security constraints all refer to ``url-pattern``s.

gregw replied:
ideally yes

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintMapping.java:
```diff
@@ -18,17 +18,29 @@
+import org.eclipse.jetty.http.pathmap.ServletPathSpec;
 import org.eclipse.jetty.util.security.Constraint;
 
 public class ConstraintMapping
```
It feels like ConstraintMapping should be a subclass of Mapping.  It could extend Mapping, but override all the array-manipulating methods to ensure that the array only ever has 0/1 elements.

gregw replied:
I'll give it a try, but it's a little ugly

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintMapping.java:
```diff
@@ -18,17 +18,29 @@
 
+    public boolean containsPathSpec(String pathSpec)
+    {
+        return _pathSpec.is(pathSpec);
```
Need null check on _pathSpec

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintMapping.java:
```diff
@@ -18,17 +18,29 @@
+
+    public boolean containsPathSpec(PathSpec pathSpec)
+    {
+        return _pathSpec.equals(pathSpec);
```
Need null check on _pathSpec.

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -67,7 +71,8 @@
     private final Set<String> _roles = new CopyOnWriteArraySet<>();
-    private final PathMap<Map<String, RoleInfo>> _constraintMap = new PathMap<>();
+
+    private final PathMappings<Map<String, RoleInfo>> _constraintMap = new PathMappings<>();
```
The change to PathMappings has already been done in jetty-10, along with a rename of _constraintMap to _constraintRoles, so make it easy on yourself and do the same rename here in jetty-9.

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -165,15 +170,7 @@ public static Constraint createConstraint(String name, String[] rolesAllowed, Em
-            }
-        }
-        return mappings;
+        return constraintMappings.stream().filter(m -> m.containsPathSpec(pathSpec)).collect(Collectors.toList());
```
For line 170, use the StringUtil.isBlank(pathSpec) test.

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -189,16 +186,8 @@ public static Constraint createConstraint(String name, String[] rolesAllowed, Em
         if (pathSpec == null || "".equals(pathSpec.trim()) || constraintMappings == null || constraintMappings.size() == 0)
```
```suggestion
        if (StringUtil.isBlank(pathSpec) || constraintMappings == null || constraintMappings.size() == 0)
```

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -165,15 +170,7 @@ public static Constraint createConstraint(String name, String[] rolesAllowed, Em
         if (pathSpec == null || "".equals(pathSpec.trim()) || constraintMappings == null || constraintMappings.size() == 0)
```
```suggestion
        if (StringUtil.isBlank(pathSpec) || constraintMappings == null || constraintMappings.size() == 0)
```

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -409,7 +409,8 @@ public void addRole(String role)
     protected void doStart() throws Exception
     {
-        _constraintMappings.stream().forEach(m -> processConstraintMapping(m));
+        _constraintMap.reset();
```
Why is this back in? It is reset in doStop()

gregw replied:
Just being safe

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -434,12 +435,12 @@ protected void doStop() throws Exception
     protected void processConstraintMapping(ConstraintMapping mapping)
     {
-        Map<String, RoleInfo> mappings = _constraintMap.get(mapping.getPathSpec());
+        Map<String, RoleInfo> mappings = _constraintMap.get(mapping.toPathSpec());
```
This is a very subtle point here: mapping.toPathSpec() will return a PathSpec, but is being implicitly coerced to a String by the jvm .... if instead you called mapping.getPathSpec() you would get a string, but it would only be non-null if it was a ServletPathSpec. So I think this needs some good javadoc here, otherwise in 6months time someone could replace the toPathSpec() with getPathSpec()!

gregw replied:
Is this the case? _constraintMap is a PathMappings, so it takes a PathSpec. No conversion to a String.

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/ConstraintSecurityHandler.java:
```diff
@@ -434,12 +435,12 @@ protected void doStop() throws Exception
         {
             mappings = new HashMap<>();
-            _constraintMap.put(mapping.getPathSpec(), mappings);
+            _constraintMap.put(mapping.toPathSpec(), mappings);
```
See previous comment!

gregw replied:
ditto

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -1350,27 +1400,23 @@ protected synchronized void updateMappings()
-            //create a map of paths to set of ServletMappings that define that mapping
-            HashMap<String, List<ServletMapping>> sms = new HashMap<>();
+            // Create a map of String paths (servlet url-pattern) to set of ServletMappings that define that mapping
+            Map<PathSpec, List<ServletMapping>> servletUrlMappings = new LinkedHashMap<>();
```
Explain why this is a LinkedHashMap.

gregw replied:
Comment added

janbartel reviewed (changes requested):
Some initial comments, I now need to do a deep dive on combining security constraints so I'll do another review when finished.

gregw committed:
Updated after review

gregw committed:
Fixed quickstart

gregw committed:
work around CDI tests

gregw committed:
better work around for owb CDI tests

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5810-bespoke-pathspec-support-alt

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-5810-bespoke-pathspec-support-alt

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Fixed test after merge

gregw committed:
updated file header

Signed-off-by: gregw <gregw@webtide.com>

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathSpec.java:
```diff
@@ -85,11 +93,165 @@
+        /**
+         * Test if the list of path specs contains a particular one.
+         *
+         * @param pathSpec the path spec
```
Is this an arbitrary type of path spec, or assumed to be a ServletPathSpec? It would be helpful to keep the language the same, as all the different path specs get a bit confusing.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -18,25 +18,29 @@
     private final int _specLength;
     private final Pattern _pattern;
 
+    private static String normalize(String regex)
```
Looking below at ServletPathSpec, should this method be called "santiize" rather than normalize? It's not prefixing "/" for example ....

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -26,17 +26,16 @@
     private final String _prefix;
     private final String _suffix;
 
     /**
```
This class should have javadoc that describes what a ServletPathSpec is:  eg prefixed with "servlet|" etc etc

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -18,25 +18,29 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class RegexPathSpec extends AbstractPathSpec
```
This class should have javadoc that describes what a RegexPathSpec is: eg prefixed by "regex|" etc etc

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -122,14 +135,19 @@ else if (servletPathSpec.charAt(0) == '*' && servletPathSpec.length() > 1)
+        return getDeclaration().equals(normalize(sanitize(pathSpec)));
+    }
+
     private static void assertValidServletPathSpec(String servletPathSpec)
```
Is the servletPathSpec arg expected to be prefixed with "servlet|" or without?

janbartel commented on jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/PathMappingsTest.java:
```diff
@@ -293,7 +293,6 @@ public void testPrecidenceVsOrdering() throws Exception
 
     @ParameterizedTest
     @ValueSource(strings = {
-        "*",
```
Why was * removed from the list?

janbartel reviewed (changes requested):
I still need to review some files - there's a lot of them. I don't see anywhere in this code where any specs are prefaced with "servlet|" or "regex|"? Where do they come from? Should this be when generating quickstart-web.xml?

gregw commented:
I think we should test this PR on jetty-10 first and only once stable consider back-porting to 9 (or even just a sponsored branch of 9)

gregw commented:
@joakime is there still a demand for this change?

joakime commented:
I'll check with the sponsor

joakime commented:
Closed in favor of Issue #7748 and associated PRs (that have already been merged)
